server/src/main/java/nl/inl/blacklab/server/datastream/DataStream.java

```diff
@@ -263,33 +263,33 @@ public DataStream entry(String key, boolean value) {
         return startEntry(key).value(value).endEntry();
     }
 
-    public DataStream attrEntry(String elementName, String attrName, String key, String value) {
-        return startAttrEntry(elementName, attrName, key).value(value).endAttrEntry();
+    public DataStream attrEntry(String elementName, String attrName, String attrValue, String value) {
```

Comment:
ok, but this should at least be documented because it is not intuitive, foo is not a key but an attribute value, and deviates from more standard xml to json serialization which would result in something like:

```
{
  "entry": {
    "key": "foo"
    "text": "bar"
  }
}
```

Reply:
The whole idea of DataStream is to avoid stuff like XML to JSON serialization, which IMO leads to verbose JSON that's not ideal to work with. I've tried to use this abstract concept of a "data stream" that can be serialized into nice XML or nice JSON. That's why a special type like contextList exists as well.

We can argue about whether the resulting XML/JSON is the right choice for the API (I think so, but would change some things if I could), and whether this implementation is ideal (it's not), but until a major rewrite we're stuck with it.

Agreed that better documentation is needed; for now I've added a comment specifically about this.

Reply:
I understand your reasons. I think it is more important to get rid of as much homebrew solutions as you can and stick to standards like jaxb, jax-ws, jsonb, cdi etc. These are there for a reason and do a splendid job. I would rather spend my time developing functionality and not worry about (sub)optimizing data structures and the like. Also such an approach garantees to some extent that my applications are future proof. And ofcourse in the course of developing contribute to the standards I and others need, so these keep being improved, maintained and promoted.

It is a rather different approach I realize that we discuss offline someday....

Reply:
I don't think our opinions are that far apart. We both agree that the current implementation isn't ideal and would be better if it used mature, well-tested libraries. I didn't do that years ago because I had to hurry and wasn't familiar with the technologies you mention, and we're stuck with the result now.

That said, I wouldn't want to compromise on the structure of the XML or JSON generated though. I think generating nice XML and nice JSON makes working with the API much easier. (not saying the current response structures are perfect btw)

If we had the time, I'm sure we could find a solution we were both happy with. Maybe some day.